src/debugAdapter.ts

```diff
@@ -2,136 +2,112 @@
+    private static readonly HEADER_FIELDSEPARATOR = /: */;
+
+    private _rawData = Buffer.allocUnsafe(0);
+    private _contentLength = -1;
+    private _isConnected: boolean = false;
+    private _debugMessageQueue: DebugProtocolMessage[] = [];
+
+    private _debugServiceSocket: Socket;
+    private _logger: Logger;
+
+    // The event that VS Code-proper will listen for.
+    private _sendMessage: EventEmitter<DebugProtocolMessage> = new EventEmitter<DebugProtocolMessage>();
+    onDidSendMessage: Event<DebugProtocolMessage> = this._sendMessage.event;
+
+    constructor(namedPipe: string, logger: Logger) {
+        this._debugServiceSocket = connect(namedPipe);
+        this._logger = logger;
+
+        this._debugServiceSocket.on("error", (e) => {
+            this._logger.writeError("Error on Debug Adapter: " + e);
+            this.dispose();
         });
 
-    // Route any output from the socket through stdout
-    debugServiceSocket.on(
-        "data",
-        (data: Buffer) => process.stdout.write(data));
-
-    // Wait for the connection to complete
-    debugServiceSocket.on(
-        "connect",
-        () => {
-            isConnected = true;
-            debugAdapterLogWriter.write("Connected to socket!\r\n\r\n");
-
-            // When data comes on stdin, route it through the socket
-            process.stdin.on(
-                "data",
-                (data: Buffer) => debugServiceSocket.write(data));
-
-            // Resume the stdin stream
-            process.stdin.resume();
-    });
-
-    // When the socket closes, end the session
-    debugServiceSocket.on(
-        "close",
-        () => {
-            debugAdapterLogWriter.write("Socket closed, shutting down.");
-            debugAdapterLogWriter.close();
-            isConnected = false;
-
-            // Close after a short delay to give the client time
-            // to finish up
-            setTimeout(() => {
-                process.exit(0);
-            }, 2000);
-        },
-    );
-
-    process.on(
-        "exit",
-        (e) => {
-            if (debugAdapterLogWriter) {
-                debugAdapterLogWriter.write("Debug adapter process is exiting...");
+        // Route any output from the socket through to VS Code.
+        this._debugServiceSocket.on("data", (data: Buffer) => this.handleData(data));
+
+        // Wait for the connection to complete.
+        this._debugServiceSocket.on("connect", () => {
+            while(this._debugMessageQueue.length) {
+                this.writeMessageToDebugAdapter(this._debugMessageQueue.shift());
             }
-        },
-    );
-}
 
-function waitForSessionFile(triesRemaining: number) {
+            this._isConnected = true;
+            this._logger.writeVerbose("Connected to socket!");
+        });
+
+        // When the socket closes, end the session.
+        this._debugServiceSocket.on("close", () => { this.dispose(); });
+    }
+
+    handleMessage(message: DebugProtocolMessage): void {
```

Comment:
Do the methods here need explicit access modifiers?

Reply:
added the public modifiers